### PR TITLE
Added more complex AIs to carps when tamed.

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request

Adds carp AI and the ability to tame carps.
Carps are now considered flying so that they don't fall through zlevels.
Generally improved movement whilst transitioning zlevels whilst buckled to a mob.

Carps can be tamed with meat, which is a random chance. Once tamed, you can ride the carp and move around with it.
You can also set it to follow you or to attack an object, similar to how corgis would.

There are some other changes to the movement of buckled mobs, which include:
- Ability to stay buckled if the buckling object is flying/floating .
- Prevents dismounting to a buckling object if it cannot pass through something. Instead, the buckling object will not be able to go any further.

Attempting to move whilst buckled to the carp will put it into a timeout which prevents AI activity for a single second, to prevent  fast movement when telling the carp to attack something. Also prevents you from freely moving about whilst the carp attacks something, requiring you to let the carp operate autonomously.

Makes dragon flying too because they're a derivative of carps.

https://cdn.discordapp.com/attachments/326831214667235328/861374496949796894/2021-07-04_23-33-35.mp4

## Why It's Good For The Game

Adds similar AI from the corgi to other simple_animal mobs, making the interactions with them slightly more advanced. You can have pet carps now!

## Changelog
:cl:
add: Added a more complex AI to carps that allow you to tell them to attack people or to follow you. These carps can be tamed with raw meat, which'll then give you control over them if you succeed in taming them. A carp will only be tamed to 1 person.
/:cl:

### CODE REVIEW

File: code/datums/ai/carp/carp_behaviors.dm [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+/datum/ai_behavior/carp_attack/perform(delta_time, datum/ai_controller/controller)
+	. = ..()
+	var/mob/living/living_pawn = controller.pawn
+	if(!istype(living_pawn) || !isturf(living_pawn.loc))
+		return
+
+	var/datum/weakref/attack_ref = controller.blackboard[BB_CARP_ATTACK_TARGET]
+	var/atom/movable/attack_target = attack_ref?.resolve()
+	if(!attack_target || !can_see(living_pawn, attack_target, length=AI_CARP_VISION_RANGE))
```
Comment (AMonkeyThatCodes) [inline code/datums/ai/carp/carp_behaviors.dm:17]: vision range should be a blackboard key. These behaviors should be generic and this entire behavior would be better off being more generic as it has no real reason to be only for carps as far as I can tell

File: code/datums/ai/carp/carp_controller.dm [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+	if(DT_PROB(5, delta_time) && (living_pawn.mobility_flags & MOBILITY_MOVE))
+		var/move_dir = pick(GLOB.alldirs)
+		living_pawn.Move(get_step(living_pawn, move_dir), move_dir)
+
+/datum/ai_controller/carp/proc/on_ridden_driver_move(atom/movable/movable_parent, mob/living/user, direction)
+	SIGNAL_HANDLER
+	PauseAi(ride_penalty_movement)
+
+/// Someone's giving us meat, let's see if we should befriend them
+/datum/ai_controller/carp/proc/on_attack_by(datum/source, obj/item/food, mob/living/attacker, params)
```
Comment (AMonkeyThatCodes) [inline code/datums/ai/carp/carp_controller.dm:75]: all this stuff thats basically taming could should be on the mob.

**Comment by TheBonded** [comment]
Can this be added to goliaths too, which you can also tame and ride?

**Comment by Watermelon914** [comment]
> Can this be added to goliaths too, which you can also tame and ride?

in a future pr, if you want to add it, then yes.

**Comment by IndieanaJones** [comment]
The dragon already has the flying element.

**Comment by Watermelon914** [comment]
> The dragon already has the flying element.

It isn't considered actual flying, they will still fall through empty spaces.

File: code/game/objects/buckling.dm [hunk 3]
```diff
@@ -120,6 +118,12 @@
 	buckled_mobs |= M
 	M.throw_alert("buckled", /atom/movable/screen/alert/buckled)
 	M.set_glide_size(glide_size)
+
+	if(can_move)
+		M.Move(loc)
+	else if (!check_loc && M.loc != loc)
+		M.forceMove(loc)
```
Comment (Kylerace) [inline code/game/objects/buckling.dm:125]: why not just forceMove() here?

**Review by Kylerace — CHANGES_REQUESTED** [review]

**Comment by Watermelon914** [comment]
for some reason I didn't get notifications you had posted this review

File: code/game/objects/buckling.dm (hunk 3 above)
Comment (Watermelon914) [reply code/game/objects/buckling.dm:125]: it unmounts the user

**Pull request merged**